aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -162,10 +171,342 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+    temp_storage.digit_count_cumsum[tidx] = digit_count_cumsum;
+  }
+  __syncthreads();
+
+  if (tidx < RADIX_DIGITS) {
+    IndexType digit_count_cumsum_left = (tidx == 0) ? 0 : temp_storage.digit_count_cumsum[tidx - 1];
+
+    // if not the last pass: update desired and ks_to_find
+    // if last pass: write out the kth value
+    if (digit_count_cumsum_left < k_to_find && k_to_find <= digit_count_cumsum) {
+      desired = at::cuda::Bitfield<Bitwise>::setBitfield(desired, tidx, current_bit, RADIX_BITS);
+      if (current_bit > 0) {
+        desires[slice_idx] = desired;
+        ks_to_find[slice_idx] = k_to_find - digit_count_cumsum_left;
+      } else {
+        kthValues[slice_idx] = TopKTypeConfig<T>::deconvert(desired);
+      }
+    }
+  }
+
+  // reset semaphores for the next pass
+  if (tidx == 0) {
+    semaphores[slice_idx] = 0;
+  }
+};
+
+template <typename T, typename IndexType, int Dim>
+void launch(
+    at::cuda::detail::TensorInfo<T, IndexType> input,
+    IndexType inputSliceSize,
+    IndexType outputSliceSize, // aka `k`
+    bool largest,
+
+    IndexType numInputSlices,
+    IndexType inputWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<T, IndexType> topK,
+    IndexType topKWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<int64_t, IndexType> indices,
+    IndexType indicesWithinSliceStride) {
+
+  // configure items_per_thread based on device architecture and input size
+  // occupancy of this kernel is limited by registers per threads
+  constexpr int REGS_PER_THREAD = 40;
+  constexpr int REGS_PER_BLOCK = REGS_PER_THREAD * BLOCK_THREADS;
+  cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+  int mpc = prop->multiProcessorCount;
+#if defined(USE_ROCM)
+  int regs_per_mp = prop->regsPerBlock;
+  int blocks_per_mp = regs_per_mp / REGS_PER_BLOCK;
+#else
+  int regs_per_mp = prop->regsPerMultiprocessor;
+  int blocks_per_mp = std::min(regs_per_mp / REGS_PER_BLOCK, prop->maxBlocksPerMultiProcessor);
+#endif
+  int items_per_thread = at::ceil_div((int64_t)(inputSliceSize * numInputSlices), (int64_t)(mpc * blocks_per_mp * BLOCK_THREADS));
+  items_per_thread = std::max(4, std::min(items_per_thread, 64)); // clamp to (4, 64)
+  int items_per_block = items_per_thread * BLOCK_THREADS;
+
+  using Bitwise = typename TopKTypeConfig<T>::RadixType;
+  int64_t blocks_per_slice = at::ceil_div((int64_t)inputSliceSize, (int64_t)items_per_block);
+  int64_t num_blocks = numInputSlices * blocks_per_slice;
+
+  // temporary storage
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+
+  auto kthValues_buffer = allocator.allocate(numInputSlices * sizeof(T));
+  T* kthValues = reinterpret_cast<T*>(kthValues_buffer.get());
+  AT_CUDA_CHECK(cudaMemsetAsync(kthValues, 0, numInputSlices * sizeof(T), c10::cuda::getCurrentCUDAStream()));
+
+  auto semaphores_buffer = allocator.allocate(numInputSlices * sizeof(int));
+  int* semaphores = reinterpret_cast<int*>(semaphores_buffer.get());
```

Comment:
In extreme case, will `blocks_per_slice` be > 2^32?

Reply:
The `atomicAdd` does not support `uint64_t`, so I changed it to `uint32_t` and check the blocks_per_slice is less than 2^32. Not perfect, but should be enough for most cases.